destination/src/main/scala/quasar/plugin/sqlserver/destination/UpsertPipe.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.plugin.sqlserver.destination
+
+import slamdata.Predef._
+
+import quasar.plugin.sqlserver._
+
+import quasar.api.push.OffsetKey
+import quasar.api.Column
+import quasar.api.resource.ResourcePath
+import quasar.connector.{DataEvent, IdBatch, MonadResourceErr}
+import quasar.connector.destination.{WriteMode => QWriteMode}
+import quasar.lib.jdbc._
+import quasar.lib.jdbc.destination.{WriteMode => JWriteMode}
+
+import cats.data.{NonEmptyList, NonEmptyVector}
+import cats.effect.{Effect, LiftIO}
+import cats.implicits._
+
+import doobie._
+import doobie.free.connection.{commit, rollback}
+import doobie.implicits._
+
+import fs2.{Pipe, Stream}
+
+import java.lang.CharSequence
+
+import org.slf4s.Logger
+
+object UpsertPipe {
+  def apply[F[_]: Effect: MonadResourceErr, A](
+      xa: Transactor[F],
+      writeMode: QWriteMode,
+      jwriteMode: JWriteMode,
+      schema: String,
+      path: ResourcePath,
+      idColumn: Option[Column[SQLServerType]],
+      inputColumns: NonEmptyList[Column[SQLServerType]],
+      logger: Logger)
+      : Pipe[F, DataEvent[CharSequence, OffsetKey.Actual[A]], OffsetKey.Actual[A]] = { events =>
+
+    val logHandler = Slf4sLogHandler(logger)
+
+    val toConnectionIO = Effect.toIOK[F] andThen LiftIO.liftK[ConnectionIO]
+
+    val hyColumns = hygienicColumns(inputColumns)
+
+    def logEvents(event: DataEvent[CharSequence, _]): F[Unit] = event match {
+      case DataEvent.Create(chunk) =>
+        trace(logger)(s"Loading chunk with size: ${chunk.size}")
+      case DataEvent.Delete(idBatch) =>
+        trace(logger)(s"Deleting ${idBatch.size} records")
+      case DataEvent.Commit(_) =>
+        trace(logger)(s"Commit")
+    }
+
+    def handleEvents(obj: Fragment, unsafeName: String)
+      : Pipe[ConnectionIO, DataEvent[CharSequence, OffsetKey.Actual[A]], Option[OffsetKey.Actual[A]]] = _ evalMap {
+
+      case DataEvent.Create(chunk) =>
+        insertChunk(logHandler)(obj, hyColumns, chunk)
+          .as(none[OffsetKey.Actual[A]])
+
+      case DataEvent.Delete(ids) if ids.size < 1 =>
+        none[OffsetKey.Actual[A]].pure[ConnectionIO]
+
+      case DataEvent.Delete(ids) => idColumn match {
+        case None =>
+          none[OffsetKey.Actual[A]].pure[ConnectionIO]
```

Comment:
This should be impossible, Idk maybe raising an error here, but this won't add any type safety anyways 🤔

Reply:
Yeah, it is a bit of an implementation detail. Might be worth moving this function to `SinkBuilder` and make it private to make it impossible to call incorrectly. It would then truly be an implementation detail as the only constructors would be `append` and `upsert`.

Reply:
Good idea :+1:

Reply:
Done